Random number generators are not and should not be differentiable

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #78106

And that's it.

Now, it's not very clear to me what are the FC BC implications of these changes.

lezcano committed:
Random number generators are not and should not be differentiable

And that's it.

Now, it's not very clear to me what are the FC BC implications of these changes.

[ghstack-poisoned]

ngimel reviewed (approved):
lgtm, but I'll wait for @albanD

soulitzer commented:
Agree about them not being differentiable, but this change in particular doesn't really improve user-facing UX around that. 

The in-place ones that had their formulas removed will now error on backward, and the error will say the backward is not implemented (rather than mention non-differentiability). For the out-of-place ones, the output will lose requires_grad (as autograd becomes a noop kernel). They will need to manually set requires_grad to true again. This is the tricky case. If they are lucky, the user may get some error about trying to call backward on a tensor that does not require grad, otherwise it is also possible that parts of their graph will silently lose gradients.

From discussion offline with @albanD, bc-breaking changes are OK in general, but we probably want to do them in the context of a larger framework. It may be good to start out with a couple examples of ops and list their current/desired behaviors and see if we can fit a couple rules around that that we can apply consistently. For example some ops like random number generators, we may want to raise an error, but for others like `==`, we may be okay with simply ignoring the requires_gradness since the output is bool anyway.

albanD commented:
I would also mention that per our current "definition" of what represents a gradient of zero (undefined, None, Tensor full of zeros, no grad). We just move from one representation of zero gradient to another. So it is hard to justify a BC-breaking change for something that doesn't change the semantic of what is computed.

albanD commented:
> Do we expect users relying on getting grad full of 0's for ops such as normal, or is losing requires_grad attribute expected?

I think both are valid. So either is ok wrt what the autograd should return.
But it is still BC-breaking change in the sense that some code that used to run will now through an error.

I'm not saying that we don't want this btw. Just that if we're making this BC-breaking change, we should be really sure that our overall story is good.

lezcano committed:
Update on "Random number generators are not and should not be differentiable"


And that's it.

Now, it's not very clear to me what are the FC BC implications of these changes.

[ghstack-poisoned]

Lezcano commented:
What is an example of code that would run without this PR and would break with this PR?

So, there are two parts to the discussion here: In-place operations and out-of-place operations. Let's start with the latter ones.

I think that in this case, this PR implements the correct behaviour. The rationale behind this is that, since this function is non-differentiable, it should be equivalent to doing things like casting to a tensor full of integers. In particular, if a tensor with `requires_grad=True` is used in the graph just as input of this operation (and not used anywhere else) we should not tell the user "sure, here's its gradient, and it happens to be zero", we should break letting the user know that there's something fundamentally wrong about their model. I believe this may uncover subtle bugs.

For the in-place versions of these operations, I do not know enough about the inner workings of the autograd and its version system, but let's see. If a user does
```python
a = torch.ones((3, 3), requires_grad=True)
a.normal_()
```
we get a 
```
RuntimeError: a leaf Variable that requires grad is being used in an in-place operation.
```
Fair enough.

If a user does
```python
a = torch.ones((3, 3), requires_grad=True)
b = a + 1
b.normal_()
b.sum().backward()
```
`b` now has a `grad_fn=<NormalBackward0>`. Now, this function does nothing, but it does fill `a.grad` with zeros. On the other hand, with this PR, this same code throws a
```
RuntimeError: derivative for aten::normal_ is not implemented
```
In my opinion, even though it is BC-breaking, the latter behaviour should be expected. The correct code imo should look like:
```python
a = torch.ones((3, 3), requires_grad=True)
b = a + 1
with torch.no_grad():
  b.normal_()
b.sum().backward()
```
This is what we do in `torch.nn.init` in any case.

I think that breaking with a noisy error regarding the derivative of `normal_` is a good thing, as it makes the user realise there's something pesky going on there. Fair enough if that's what they'd like to do, but if so they should be explicit about their intents using `torch.no_grad()`, and if they do so nothing will break.

Now, I understand that this is a big BC-breaking change, and I'd understand if we may not want to do this, but I think this would be good. Also, I think that the error of "derivative not implemented" is not too bad, but I do agree that it'd be even better if we could throw an error of the form "`aten::normal_` is not differentiable".

lezcano committed:
Update on "Random number generators are not and should not be differentiable"


And that's it.

Now, it's not very clear to me what are the FC BC implications of these changes.

[ghstack-poisoned]

Lezcano commented on test/test_nn.py:
```diff
@@ -13191,7 +13191,8 @@ def _test_GroupNorm_general(self, device, dtype=torch.float):
                 gn = nn.GroupNorm(g, shape[1])
 
     def _test_GroupNorm_cuda_half(self):
-        input = torch.zeros(2, 4, 3, 2, requires_grad=True).cuda().half().random_(1, 10)
```
Example of an incorrect use of `random_` detected by this PR. I am not 100% sure whether the previous code has  performance implications (e.g. saving unnecessary tensors or doing extra work in the `backward`), but it is clearly an anti-pattern. Initialisation of tensors should not propagate gradients.

Lezcano commented:
I think you meant `torch.normal(a)` (or more explicitly `torch.normal(mean=a)`). In this case, note that this PR moves the behaviour of your snippet of code to be the same as that of
```python
a = torch.tensor(1., requires_grad=True)
b = torch.tensor(1., requires_grad=True)
c = torch.ones_like(a) + b
c.backward()
```
which seems more than reasonable.

About the error nodes, I am a bit ambivalent. On the one hand, I think it could be good to make the users more aware of what they're doing and make them surround their sampling by a `torch.no_grad()`. On the other hand, it makes sense to me, as mentioned above, that, from an API perspective, `torch.ones_like` and `torch.normal` behave similarly, even though the first one is differentiable (it's a constant function) and the latter one is not.

I tend to lean more for the latter option (consistent API) than the former, but what do you guys think?

Also, I'm all in favour of improving the error message for the in-place variants.

lezcano committed:
Update on "Random number generators are not and should not be differentiable"


And that's it.

Now, it's not very clear to me what are the FC BC implications of these changes.

[ghstack-poisoned]

lezcano committed:
Update on "Random number generators are not and should not be differentiable"


And that's it.

Now, it's not very clear to me what are the FC BC implications of these changes.

[ghstack-poisoned]

lezcano committed:
Update on "Random number generators are not and should not be differentiable"


And that's it.

Now, it's not very clear to me what are the FC BC implications of these changes.

[ghstack-poisoned]

lezcano committed:
Update on "Random number generators are not and should not be differentiable"


And that's it.

Now, it's not very clear to me what are the FC BC implications of these changes.

[ghstack-poisoned]

Lezcano commented:
I pulled it out of the other stack to unlock it, as per the offline discussion.

lezcano committed:
Update on "Random number generators are not and should not be differentiable"


And that's it.

Now, it's not very clear to me what are the FC BC implications of these changes.

[ghstack-poisoned]

lezcano committed:
Update on "Random number generators are not and should not be differentiable"


And that's it.

Now, it's not very clear to me what are the FC BC implications of these changes.

[ghstack-poisoned]

lezcano committed:
Update on "Random number generators are not and should not be differentiable"


And that's it.

Now, it's not very clear to me what are the FC BC implications of these changes.

[ghstack-poisoned]

lezcano committed:
Update on "Random number generators are not and should not be differentiable"


And that's it.

Now, it's not very clear to me what are the FC BC implications of these changes.

[ghstack-poisoned]

lezcano committed:
Update on "Random number generators are not and should not be differentiable"


And that's it.

Now, it's not very clear to me what are the FC BC implications of these changes.

[ghstack-poisoned]

Lezcano commented:
cc @fritzo

lezcano committed:
Update on "Random number generators are not and should not be differentiable"


And that's it.

Now, it's not very clear to me what are the FC BC implications of these changes.

[ghstack-poisoned]

Lezcano commented:
Closing this one for now. We'll come back to this one once we do a complete overhaul of non-differentiability all across PyTorch.
